File: backend/app/schemas/dataset.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict


class TopValue(BaseModel):
    value: object
    count: int


class ColumnProfile(BaseModel):
    name: str
    dtype: str
    null_count: int
    null_percentage: float = 0.0
    # Numeric columns only.
    min: float | int | None = None
    max: float | int | None = None
    mean: float | None = None
    outlier_count: int | None = None
    # Categorical (string) columns only.
    top_values: list[TopValue] | None = None


class Correlation(BaseModel):
    column_a: str
    column_b: str
    correlation: float


class DatasetRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    original_filename: str
    content_type: str
    size_bytes: int
    status: str
    created_at: datetime

    # Populated once profiling completes (status == "profiled"); null while
    # "uploaded" or if it failed ("profiling_failed").
    row_count: int | None = None
    column_count: int | None = None
    columns_profile: list[ColumnProfile] | None = None
    # Data quality (V2) — see app.analytics.profiling.
    duplicate_row_count: int | None = None
    data_quality_score: float | None = None
    correlations: list[Correlation] | None = None

File: frontend/src/types/dataset.ts
export interface TopValue {
  value: unknown
  count: number
}

export interface ColumnProfile {
  name: string
  dtype: string
  null_count: number
  null_percentage: number
  min: number | null
  max: number | null
  mean: number | null
  outlier_count: number | null
  top_values: TopValue[] | null
}

export interface Correlation {
  column_a: string
  column_b: string
  correlation: number
}

export interface Dataset {
  id: string
  original_filename: string
  content_type: string
  size_bytes: number
  status: string
  created_at: string
  row_count: number | null
  column_count: number | null
  columns_profile: ColumnProfile[] | null
  duplicate_row_count: number | null
  data_quality_score: number | null
  correlations: Correlation[] | null
}

export type AggregationOp = 'sum' | 'avg' | 'min' | 'max' | 'count'
export type FilterOp = 'eq' | 'ne' | 'gt' | 'gte' | 'lt' | 'lte' | 'contains'

export interface Aggregation {
  op: AggregationOp
  column?: string
  alias?: string
}

export interface QueryFilter {
  column: string
  op: FilterOp
  value: string | number | boolean
}

export interface DatasetQueryRequest {
  group_by?: string[]
  aggregations?: Aggregation[]
  filters?: QueryFilter[]
  sort_by?: string
  sort_desc?: boolean
  limit?: number
}

export interface DatasetQueryResult {
  columns: string[]
  rows: unknown[][]
  row_count: number
  truncated: boolean
}
